README.rst
.. image:: https://raw.github.com/jvllmr/mixer/develop/docs/_static/logo.png
    :width: 100px


.. important::

   This is a fork of https://github.com/klen/mixer that aims to maintain integrations with the newest library versions.
   It also improves Developer Experience by being fully typed.


The **Mixer** is a helper to generate instances of Django or SQLAlchemy models.
It's useful for testing and fixture replacement. Fast and convenient test-data
generation.

Mixer supports:

* Django_;
* SQLAlchemy_;
* Flask-SQLAlchemy_;
* Peewee_;
* Pony_;
* Mongoengine_;
* Marshmallow_;
* Custom schemes;

.. _badges:

.. image:: https://github.com/jvllmr/mixer/workflows/tests/badge.svg?style=flat-square
    :target: https://github.com/jvllmr/mixer/actions
    :alt: Tests Status

.. image:: http://img.shields.io/pypi/v/mixer2.svg?style=flat-square
    :target: https://pypi.python.org/pypi/mixer2
    :alt: Version

.. image:: http://img.shields.io/pypi/dm/mixer2.svg?style=flat-square
    :target: https://pypi.python.org/pypi/mixer2
    :alt: Downloads

.. image:: http://img.shields.io/pypi/l/mixer2.svg?style=flat-square
    :target: https://pypi.python.org/pypi/mixer2
    :alt: License

.. _documentation:


**Docs are available at https://mixer.readthedocs.org/. Pull requests with
documentation enhancements and/or fixes are awesome and most welcome.**

Описание на русском языке: http://klen.github.io/mixer.html

.. important::

   From version 6.2 the Mixer library doesn't support Python 2.
   The latest version with python<3 support is mixer 6.1.3


.. _contents:

.. contents::


Requirements
=============

- Python 3.8+
- Django (3.0, 3.1) for Django ORM support;
- Flask-SQLALchemy for SQLAlchemy ORM support and integration as Flask application;
- Faker >= 0.7.3
- Mongoengine for Mongoengine ODM support;
- SQLAlchemy for SQLAlchemy ORM support;
- Peewee ORM support;


Installation
=============

**Mixer** should be installed using pip: ::

    pip install mixer2


Usage
=====

 |   By default Mixer tries to generate fake (human-friendly) data.
 |   If you want to randomize the generated values initialize the Mixer
 |   by manual: Mixer(fake=False)


 |   By default Mixer saves the generated objects in a database. If you want to disable
 |   this, initialize the Mixer by manual like Mixer(commit=False)


Django workflow
---------------
Quick example:

.. code-block:: python

    from mixer2.backend.django import mixer
    from customapp.models import User, UserMessage

    # Generate a random user
    user = mixer.blend(User)

    # Generate an UserMessage
    message = mixer.blend(UserMessage, user=user)

    # Generate an UserMessage and an User. Set username for generated user to 'testname'.
    message = mixer.blend(UserMessage, user__username='testname')

    # Generate SomeModel from SomeApp and select FK or M2M values from db
    some = mixer.blend('someapp.somemodel', somerelation=mixer.SELECT)

    # Generate SomeModel from SomeApp and force a value of money field from default to random
    some = mixer.blend('someapp.somemodel', money=mixer.RANDOM)

    # Generate SomeModel from SomeApp and skip the generation of money field
    some = mixer.blend('someapp.somemodel', money=mixer.SKIP)

    # Generate 5 SomeModel's instances and take company field's values from custom generator
    some_models = mixer.cycle(5).blend('somemodel', company=(name for name in company_names))


Flask, Flask-SQLAlchemy
-----------------------
Quick example:

.. code-block:: python

    from mixer2.backend.flask import mixer
    from models import User, UserMessage

    mixer.init_app(self.app)

    # Generate a random user
    user = mixer.blend(User)

    # Generate an userMessage
    message = mixer.blend(UserMessage, user=user)

    # Generate an UserMessage and an User. Set username for generated user to 'testname'.
    message = mixer.blend(UserMessage, user__username='testname')

    # Generate SomeModel and select FK or M2M values from db
    some = mixer.blend('project.models.SomeModel', somerelation=mixer.SELECT)

    # Generate SomeModel from SomeApp and force a value of money field from default to random
    some = mixer.blend('project.models.SomeModel', money=mixer.RANDOM)

    # Generate SomeModel from SomeApp and skip the generation of money field
    some = mixer.blend('project.models.SomeModel', money=mixer.SKIP)

    # Generate 5 SomeModel's instances and take company field's values from custom generator
    some_models = mixer.cycle(5).blend('project.models.SomeModel', company=(company for company in companies))


Support for Flask-SQLAlchemy models that have `__init__` arguments
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For support this scheme, just create your own mixer class, like this:

.. code-block:: python

    from mixer2.backend.sqlalchemy import Mixer

    class MyOwnMixer(Mixer):

        def populate_target(self, values):
            target = self.__scheme(**values)
            return target

    mixer = MyOwnMixer()


SQLAlchemy workflow
-------------------

Example of initialization:

.. code-block:: python

    from mixer2.backend.sqlalchemy import Mixer

    ENGINE = create_engine('sqlite:///:memory:')
    BASE = declarative_base()
    SESSION = sessionmaker(bind=ENGINE)

    mixer = Mixer(session=SESSION(), commit=True)
    role = mixer.blend('package.models.Role')


Also, see `Flask`_, `Flask-SQLAlchemy`_.


Mongoengine workflow
--------------------

Example usage:

.. code-block:: python

    from mixer2.backend.mongoengine import mixer

    class User(Document):
        created_at = DateTimeField(default=datetime.datetime.now)
        email = EmailField(required=True)
        first_name = StringField(max_length=50)
        last_name = StringField(max_length=50)
        username = StringField(max_length=50)

    class Post(Document):
        title = StringField(max_length=120, required=True)
        author = ReferenceField(User)
        tags = ListField(StringField(max_length=30))

    post = mixer.blend(Post, author__username='foo')

Marshmallow workflow
--------------------

Example usage:

.. code-block:: python

    from mixer2.backend.marshmallow import mixer
    import marshmallow as ma

    class User(ma.Schema):
        created_at = ma.fields.DateTime(required=True)
        email = ma.fields.Email(required=True)
        first_name = ma.fields.String(required=True)
        last_name = ma.fields.String(required=True)
        username = ma.fields.String(required=True)

    class Post(ma.Schema):
        title = ma.fields.String(required=True)
        author = ma.fields.Nested(User, required=True)

    post = mixer.blend(Post, author__username='foo')


Common usage
------------
Quick example:

.. code-block:: python

    from mixer2.main import mixer

    class Test:
        one = int
        two = int
        name = str

    class Scheme:
        name = str
        money = int
        male = bool
        prop = Test

    scheme = mixer.blend(Scheme, prop__one=1)

Type annotations, typing.TypedDict and python dataclasses
------------
Example:

.. code-block:: python

    from mixer2.backend.annotated import mixer
    from typing import TypedDict

    class Test:
        one: int
        two: int
        name: str

    class Scheme(TypedDict):
        name: str
        money: int
        male: bool
        prop: Test

    scheme = mixer.blend(Scheme, name="John", male=True) # {"name": "John", "male": True, ...}

It works the same with python dataclasses.

DB commits
----------

By default 'django', 'flask', 'mongoengine' backends tries to save objects in
database. For preventing this behavior init `mixer` manually:

.. code-block:: python

    from mixer2.backend.django import Mixer

    mixer = Mixer(commit=False)


Or you can temporary switch context use the mixer as context manager:

.. code-block:: python

    from mixer2.backend.django import mixer

    # Will be save to db
    user1 = mixer.blend('auth.user')

    # Will not be save to db
    with mixer.ctx(commit=False):
        user2 = mixer.blend('auth.user')


.. _custom:

Custom fields
-------------

The mixer allows you to define generators for fields by manually.
Quick example:

.. code-block:: python

        from mixer2.main import mixer

        class Test:
            id = int
            name = str

        mixer.register(Test,
            name=lambda: 'John',
            id=lambda: str(mixer.faker.small_positive_integer())
        )

        test = mixer.blend(Test)
        test.name == 'John'
        isinstance(test.id, str)

        # You could pinned just a value to field
        mixer.register(Test, name='Just John')
        test = mixer.blend(Test)
        test.name == 'Just John'

Also, you can make your own factory for field types:

.. code-block:: python

    from mixer2.backend.django import Mixer, GenFactory

    def get_func(*args, **kwargs):
        return "Always same"

    class MyFactory(GenFactory):
        generators = {
            models.CharField: get_func
        }

    mixer = Mixer(factory=MyFactory)

Middlewares
-----------

You can add middleware layers to process generation:

.. code-block:: python

    from mixer2.backend.django import mixer

    # Register middleware to model
    @mixer.middleware('auth.user')
    def encrypt_password(user):
        user.set_password('test')
        return user

You can add several middlewares. Each middleware should get one argument
(generated value) and return them.

It's also possible to unregister a middleware:

.. code-block:: python

    mixer.unregister_middleware(encrypt_password)


Locales
-------

By default mixer uses 'en' locale. You could switch mixer default locale by
creating your own mixer:

.. code-block:: python

    from mixer2.backend.django import Mixer

    mixer = Mixer(locale='it')
    mixer.faker.name()          ## u'Acchisio Conte'

At any time you could switch mixer current locale:

.. code-block:: python

    mixer.faker.locale = 'cz'
    mixer.faker.name()          ## u'Miloslava Urbanov\xe1 CSc.'

    mixer.faker.locale = 'en'
    mixer.faker.name()          ## u'John Black'

    # Use the mixer context manager
    mixer.faker.phone()         ## u'1-438-238-1116'
    with mixer.ctx(locale='fr'):
        mixer.faker.phone()     ## u'08 64 92 11 79'

    mixer.faker.phone()         ## u'1-438-238-1116'

.. _bugtracker:

Bug tracker
===========

If you have any suggestions, bug reports or
annoyances please report them to the issue tracker
at https://github.com/klen/mixer/issues


Contributing
============

Development of mixer happens at Github: https://github.com/klen/mixer


Contributors
=============

* Antoine Bertin            (https://github.com/Diaoul)
* Benjamin Port             (https://github.com/bport)
* Dmitriy Moseev            (https://github.com/DmitriyMoseev)
* Eelke Hermens             (https://github.com/eelkeh)
* Esteban J. G. Gabancho    (https://github.com/egabancho)
* Felix Dreissig            (https://github.com/F30)
* Illia Volochii            (https://github.com/illia-v)
* Jannis                    (https://github.com/jnns)
* Kirill Pavlov             (https://github.com/pavlov99)
* Kwok-kuen Cheung          (https://github.com/cheungpat)
* Mahdi Yusuf               (https://github.com/myusuf3)
* Marek Baczyński           (https://github.com/imbaczek)
* Marigold                  (https://github.com/Marigold)
* Matt Caldwell             (https://github.com/mattcaldwell)
* Mikhail Porokhovnichenko  (https://github.com/marazmiki)
* Skylar Saveland           (https://github.com/skyl)
* Suriya Subramanian        (https://github.com/suriya)
* Gram                      (https://github.com/orsinium)
* Joshua                    (https://github.com/jomasti)
* Lucas Rangel Cezimbra     (https://github.com/lucasrcezimbra)
* avi-pomicell              (https://github.com/avi-pomicell)
* Jochen Brissier           (https://github.com/jbrissier)


License
========

Licensed under a `BSD license`_.


.. _links:

.. _Django: http://djangoproject.com/
.. _Flask: https://flask.palletsprojects.com/en/1.1.x/
.. _Flask-SQLAlchemy: http://flask-sqlalchemy.pocoo.org/
.. _SQLAlchemy: http://www.sqlalchemy.org/
.. _Marshmallow: http://marshmallow.readthedocs.io/en/latest/
.. _Mongoengine: http://mongoengine.org/
.. _Peewee: http://peewee.readthedocs.org/en/latest/
.. _Pony: http://ponyorm.com/
.. _klen: http://klen.github.io
.. _BSD license: http://www.linfo.org/bsdlicense.html


Changelog
2022-07-10
	* Add support for type annotations, typing.TypedDict and dataclasses

2022-03-23
	* Django: Support for UUIDField

2021-10-13
	* Support latest libraries
	* Marshmallow: support Dict and Raw fields

2020-12-30

	* Drop Python2 Support
	* Update Faker to 4.0.0

2018-09-25

	* mixer.unregister_middleware

2017-11-06  marazmiki

	* Added Django 2.0 support for Python 3.6;
	* Added Django 1.10 support for Python 3.6;
	* Added Django 1.11 support for Python 3.5+.
	* Droped support for Django 1.9-

2016-11-16  horneds

	* Basic Marshmallow support
	* Faker 0.7.3
	* Add Django 1.10 support
	* Drop Django 1.7 support

2016-03-18  horneds

	* fake-factory 5.0.7
	* fix deprication warnings (django)
	* Add Django 1.9 support
	* Drop Django 1.6 support

2015-04-14  horneds

	* Version 5.1.0
	* Drop support for Django 1.5.*
	* Add support for Django 1.8
	* Add support for mongoengine 0.9.0

2015-01-23  horneds

	* Version 5.0.0
	* fake-factory (Faker) integrated to Mixer
	* Support locales
	* WARNING: API has been changed (mixer.G, mixer.F has been removed,
	use mixer.faker insted)

2014-11-18  horneds

	* Support timezones in Django
	* Fix installation in python 2.6

2014-09-29  horneds

	* Fix SQLAlchemy <= 0.8 compatibility
	* Version 4.9.3

2014-09-15  horneds

	* Drop support Django 1.4
	* Add support Django 1.7

2014-08-18  horneds

	* Version 4.8.1
	* Fix sqlalchemy using default GenFactory (c) cheungpat

2014-08-04  horneds

	* Version 4.8.0
	* Added gen_timedelta and gen_datetime_range to generators

2014-08-31  horneds

	* Version 4.7.5
	* Support EmailField in YADM

2014-07-10  horneds

	* Version 4.7.4
	* Fix phone numbers generation

2014-07-09  horneds

	* Version 4.7.2
	* Base support for custom fields

2014-06-19  horneds

	* The package's exceptions is more informative.

2014-06-17  horneds

	* Version 4.6.0
	* Provided guard and select for sqlalchemy. Added mixer.reload.
	* Better Django support. Now Mixer tracks field value's type and
	update it if needed

2014-06-13  horneds

	* Version 4.5.3
	* Add support for django fields: FilePathField, GenericIPAddressField, PositiveSmallIntegerField

2014-06-07  horneds

	* Version 4.5.2
	* Fix sqlalchemy many to many relationships (c) Antoine Bertin

2014-05-23  horneds

	* Version 4.5.1
	* More information from errors

2014-05-19  horneds

	* Version 4.4.0
	* Added mixer.G.get_percent and mixer.G.get_percent_decimal

2014-05-16  horneds

	* Version 4.3.2
	* Add support for Django o2o field.

2014-05-14  horneds

	* Version 4.3.1
	* Fix get_positive_decimal

2014-05-13  horneds

	* Version 4.3.0
	* Improve lorem generation
	* Added slug faker

2014-05-07  horneds

	* Version 4.0.7
	* Minor fixes

2014-05-05  horneds

	* Version 4.0.5
	* Respect max_length for django.TextField

2014-04-30  horneds

	* Version 4.0.4
	* Add beta support for Peewee and Pony ORM

2014-04-16  horneds

	* Version 3.0.0
	* Django: ManyToManyField is not generated by default
	* Change mixer.register syntax (See README)
	* Instead postprocess Mixer now supports middleware decorators (See README)

2014-02-10  horneds

	* Version 2.3.0
	* Added beta support for guards
	    # Works like "get or create user with username='root'"
	    root = mixer.guard(username='root').blend('auth.user')

2014-02-09  horneds

	* Version 2.2.0
	* Added 'silence' param to mixer initialization
	    with mixer.ctx(silence=True):
	        # No errors will be raised
		user = mixer.blend('auth.User', username='mike')
		null = mixer.blend('auth.User', username='mike')


2014-01-03  horneds

	* Version 2.1.0
	* Updated support for django content_types

2013-12-31  horneds

	* Version 2.0.0
	* 'mixer.random', 'mixer.fake', 'mixer.select', 'mixer.mix',
	'mixer.f', 'mixer.g' -- are depricated use them in UPPERCASE ('mixer.RANDOM' and etc)
	* Added 'mixer.SKIP' attribute for skipping a value generation

2013-12-27  horneds

	* Support context (redefine mixer params as context)
	* Version 1.6.0

2013-11-25  horneds

	* Version 1.5.1
	* Fixed builtin types generation

2013-11-22  horneds

	* Version 1.5.0
	* Django ContentTypes integration was be upgraded

2013-11-21  horneds

	* Version 1.4.0
	* Now postprocess working before save

2013-11-15  horneds

	* Version 1.3.0
	* Added postprocess param to mixer.register method
	* Added gen_simple_username faker

2013-09-22  horneds

	* Version 1.1.6
	* Bugfix release

2013-09-18  horneds

	* Version 1.1.4
	* Fix MongoEngine backend (c) Kirill Pavlov
	* By default mongoengine save models in database

2013-09-11  horneds

	* Version 1.1.2
	* SQLAlchemy bugfix

2013-09-06  horneds

	* Version 1.1.1
	* `mixer.register` could pinned just a values to field

2013-09-05  horneds

	* Version 1.0.0
	* Added `mixer.register` method (register custom generators for fields)

2013-09-04  horneds

	* Version 0.11.3
	* Fixed Django m2m

2013-08-21  horneds

	* Version 0.11.2
	* Minor changes

2013-07-31  horneds

	* Version 0.11.1
	* Added FileField and ImageField support to django backend;
	* Supports Flask-SQLAlchemy for python 3;
	* [Django] Supports model creation only by model name. For now you can
	  use three methods:

	    1) mixer.blend(SomeModel) # by class
	    2) mixer.blend('someapp.somemodel') # by 'app.model' string
	    3) mixer.blend('somemodel') # short form for unique model names

2013-07-15  horneds

	* Version 0.9.1
	* Allow mixer.sequence to get a few argument and make generators from them
	* Fix bug with 'mixer.fake' attribute;

2013-07-05  horneds

	* Version 0.8.1
	* Change float's generator (limits set to -99999 -- 99999)

2013-06-26  horneds

	* Version 0.8.0
	* Added automatic backend

2013-06-22  horneds

	* Version 0.7.0
	* Minor changes

2013-06-21  horneds

	* Version 0.6.4
	* Support django.models.URLField

2013-06-20  horneds

	* Version 0.6.3
	* Now `mixer.random` can accept a choices and return random value
	  http://mixer.readthedocs.org/en/latest/api.html#force-a-random-values

2013-06-19  horneds

	* Version 0.6.2

mixer/backend/django.py
""" Django support. """

import datetime as dt
import decimal
from os import path
from types import GeneratorType

from django.apps import apps
from django.conf import settings
from django.contrib.contenttypes.fields import (
    GenericForeignKey,
    GenericRelation,
)
from django.contrib.contenttypes.models import ContentType
from django.core.files.base import ContentFile
from django.core.validators import validate_ipv4_address, validate_ipv6_address
from django.db import models

from .. import _compat as _
from .. import mix_types as t
from ..main import SKIP_VALUE
from ..main import GenFactory as BaseFactory
from ..main import Mixer as BaseMixer
from ..main import TypeMixer as BaseTypeMixer
from ..main import TypeMixerMeta as BaseTypeMixerMeta
from ..main import faker, partial

get_contentfile = ContentFile

MOCK_FILE = path.abspath(
    path.join(path.dirname(path.dirname(__file__)), "resources", "file.txt")
)
MOCK_IMAGE = path.abspath(
    path.join(path.dirname(path.dirname(__file__)), "resources", "image.gif")
)


class UTCZone(dt.tzinfo):

    """Implement UTC timezone."""

    utcoffset = dst = lambda s, d: dt.timedelta(0)
    tzname = lambda s, d: "UTC"


UTC = UTCZone()


def get_file(filepath=MOCK_FILE, **kwargs):
    """Generate a content file.

    :return ContentFile:

    """
    with open(filepath, "rb") as f:
        name = path.basename(filepath)
        return get_contentfile(f.read(), name)


def get_image(filepath=MOCK_IMAGE):
    """Generate a content image.

    :return ContentFile:

    """
    return get_file(filepath)


def get_relation(_scheme=None, _typemixer=None, **params):
    """Function description."""
    scheme = _scheme.related_model

    if scheme is ContentType:
        choices = [m for m in apps.get_models() if m is not ContentType]
        return ContentType.objects.get_for_model(faker.random_element(choices))

    return TypeMixer(
        scheme,
        mixer=_typemixer._TypeMixer__mixer,
        factory=_typemixer._TypeMixer__factory,
        fake=_typemixer._TypeMixer__fake,
    ).blend(**params)


def get_datetime(**params):
    """Support Django TZ support."""
    return faker.date_time(tzinfo=UTC if settings.USE_TZ else None)


class GenFactory(BaseFactory):

    """Map a django classes to simple types."""

    types = {
        (models.AutoField, models.PositiveIntegerField): t.PositiveInteger,
        models.BigIntegerField: t.BigInteger,
        models.BooleanField: bool,
        (models.CharField, models.SlugField): str,
        models.DateField: dt.date,
        models.DecimalField: decimal.Decimal,
        models.EmailField: t.EmailString,
        models.FloatField: float,
        models.GenericIPAddressField: t.IPString,
        models.IPAddressField: t.IP4String,
        models.IntegerField: int,
        models.PositiveSmallIntegerField: t.PositiveSmallInteger,
        models.SmallIntegerField: t.SmallInteger,
        models.TextField: t.Text,
        models.TimeField: dt.time,
        models.URLField: t.URL,
        models.UUIDField: t.UUID,
    }

    generators = {
        models.BinaryField: faker.pybytes,
        models.DateTimeField: get_datetime,
        models.FileField: get_file,
        models.FilePathField: lambda: MOCK_FILE,
        models.ForeignKey: get_relation,
        models.ImageField: get_image,
        models.ManyToManyField: get_relation,
        models.OneToOneField: get_relation,
    }


class TypeMixerMeta(BaseTypeMixerMeta):

    """Load django models from strings."""

    def __new__(mcs, name, bases, params):
        """Associate Scheme with Django models.

        Cache Django models.

        :return mixer.backend.django.TypeMixer: A generated class.

        """
        params["models_cache"] = dict()
        cls = super().__new__(mcs, name, bases, params)
        return cls

    def __load_cls(cls, cls_type):
        if isinstance(cls_type, _.string_types):
            if "." in cls_type:
                app_label, model_name = cls_type.split(".")
                return apps.get_model(app_label, model_name)

            else:
                try:
                    if cls_type not in cls.models_cache:
                        cls.__update_cache()

                    return cls.models_cache[cls_type]

                except KeyError:
                    raise ValueError('Model "%s" not found.' % cls_type)

        return cls_type

    def __update_cache(cls):
        """Update apps cache for Django < 1.7."""
        for app in apps.all_models:
            for name, model in apps.all_models[app].items():
                cls.models_cache[name] = model


class TypeMixer(_.with_metaclass(TypeMixerMeta, BaseTypeMixer)):

    """TypeMixer for Django."""

    __metaclass__ = TypeMixerMeta

    factory = GenFactory

    def postprocess(self, target, postprocess_values):
        """Fill postprocess_values."""
        for name, deffered in postprocess_values:
            if not isinstance(deffered.scheme, GenericForeignKey):
                continue

            name, value = self._get_value(name, deffered.value)
            setattr(target, name, value)

        if self.__mixer:
            target = self.__mixer.postprocess(target)

        for name, deffered in postprocess_values:
            if isinstance(deffered.scheme, GenericForeignKey) or not target.pk:
                continue

            name, value = self._get_value(name, deffered.value)
            # # if the value ist the skip value
            # # stop further processing
            if value == SKIP_VALUE:
                continue
            # # If the ManyToMany relation has an intermediary model,
            # # the add and remove methods do not exist.
            through = deffered.scheme.remote_field.through

            if not through._meta.auto_created and self.__mixer:
                self.__mixer.blend(
                    through,
                    **{
                        deffered.scheme.m2m_field_name(): target,
                        deffered.scheme.m2m_reverse_field_name(): value,
                    },
                )
                continue

            if not isinstance(value, (list, tuple)):
                value = [value]

            getattr(target, name).set(value)

        return target

    def get_value(self, name, value):
        """Set value to generated instance.

        :return : None or (name, value) for later use

        """
        field = self.__fields.get(name)
        if field:
            if field.scheme in self.__scheme._meta.local_many_to_many or isinstance(
                field.scheme, GenericForeignKey
            ):
                return name, t._Deffered(value, field.scheme)

            return self._get_value(name, value, field)

        return super().get_value(name, value)

    def _get_value(self, name, value, field=None):
        if isinstance(value, GeneratorType):
            return self._get_value(name, next(value), field)

        if not isinstance(value, t.Mix) and value is not SKIP_VALUE:
            if callable(value):
                return self._get_value(name, value(), field)

            if field and not isinstance(field.scheme, models.ForeignKey):
                value = field.scheme.to_python(value)

        return name, value

    def gen_select(self, field_name, select):
        """Select exists value from database.

        :param field_name: Name of field for generation.

        :return : None or (name, value) for later use

        """
        if field_name not in self.__fields:
            return field_name, None

        try:
            field = self.__fields[field_name]
            return (
                field.name,
                field.scheme.remote_field.model.objects.filter(
                    **select.params
                ).order_by("?")[0],
            )

        except Exception:
            raise Exception(f"Cannot find a value for the field: '{field_name}'")

    def gen_field(self, field):
        """Generate value by field.

        :param relation: Instance of :class:`Field`

        :return : None or (name, value) for later use

        """
        if isinstance(field.scheme, GenericForeignKey):
            return field.name, SKIP_VALUE

        if field.params and not field.scheme:
            raise ValueError("Invalid relation %s" % field.name)

        return super().gen_field(field)

    def make_fabric(self, field, fname=None, fake=False, kwargs=None):
        """Make a fabric for field.

        :param field: A mixer field
        :param fname: Field name
        :param fake: Force fake data

        :return function:

        """
        kwargs = {} if kwargs is None else kwargs

        fcls = type(field)
        stype = self.__factory.cls_to_simple(fcls)

        if fcls is models.CommaSeparatedIntegerField:
            return partial(
                faker.random_sample, range(0, field.max_length), length=field.max_length
            )

        if field and field.choices:
            try:
                choices, _ = list(zip(*field.choices))
                return partial(faker.random_element, choices)
            except ValueError:
                pass

        if stype in (str, t.Text):
            fab = super().make_fabric(fcls, field_name=fname, fake=fake, kwargs=kwargs)
            return lambda: fab()[: field.max_length]

        if stype is decimal.Decimal:
            kwargs["left_digits"] = field.max_digits - field.decimal_places
            kwargs["right_digits"] = field.decimal_places

        elif stype is t.IPString:
            # Hack for support Django 1.4/1.5
            protocol = getattr(field, "protocol", None)
            if not protocol:
                validator = field.default_validators[0]
                protocol = "both"
                if validator is validate_ipv4_address:
                    protocol = "ipv4"
                elif validator is validate_ipv6_address:
                    protocol = "ipv6"

            # protocol matching is case insensitive
            # default address is either IPv4 or IPv6
            kwargs["protocol"] = protocol.lower()

        elif isinstance(field, models.fields.related.RelatedField):
            kwargs.update({"_typemixer": self, "_scheme": field})

        return super().make_fabric(fcls, field_name=fname, fake=fake, kwargs=kwargs)

    @staticmethod
    def is_unique(field):
        """Return True is field's value should be a unique.

        :return bool:

        """
        return field.scheme.unique

    @staticmethod
    def is_required(field):
        """Return True is field's value should be defined.

        :return bool:

        """
        if field.params:
            return True

        if field.scheme.has_default() or field.scheme.null and field.scheme.blank:
            return False

        if field.scheme.auto_created:
            return False

        if isinstance(field.scheme, models.ManyToManyField):
            return False

        if isinstance(field.scheme, GenericRelation):
            return False

        return True

    def guard(self, *args, **kwargs):
        """Look objects in database.

        :returns: A finded object or False

        """
        qs = self.__scheme.objects.filter(*args, **kwargs)
        count = qs.count()

        if count == 1:
            return qs.get()

        if count:
            return list(qs)

        return False

    def reload(self, obj):
        """Reload object from database."""
        if not obj.pk:
            raise ValueError("Cannot load the object: %s" % obj)
        return self.__scheme._default_manager.get(pk=obj.pk)

    def __load_fields(self):
        private_fields = getattr(self.__scheme._meta, "private_fields", [])

        for field in private_fields:
            yield field.name, t.Field(field, field.name)

        for field in self.__scheme._meta.fields:
            if (
                isinstance(field, models.AutoField)
                and self.__mixer
                and self.__mixer.params.get("commit")
            ):
                continue

            yield field.name, t.Field(field, field.name)

        for field in self.__scheme._meta.local_many_to_many:
            yield field.name, t.Field(field, field.name)


class Mixer(BaseMixer):

    """Integration with Django."""

    type_mixer_cls = TypeMixer

    def __init__(self, commit=True, **params):
        """Initialize Mixer instance.

        :param commit: (True) Save object to database.

        """
        super().__init__(**params)
        self.params["commit"] = commit

    def postprocess(self, target):
        """Save objects in db.

        :return value: A generated value

        """
        if self.params.get("commit"):
            target.save()

        return target


# Default mixer
mixer = Mixer()

# pylama:ignore=E1120
